archival: Improve rw-fence

Lazin opened the pull request:
This PR fixes few bugs in the `ntp_archiver`.
- The fence is acquired without the mutex in the main upload loop. Because of that sometimes uploads could be rejected if something is happening in parallel.
- The fence is acquired before the mutex units in 'find_reupload_candidate'.
- The 'reset scrubbing metadata' call is performed without the fence or acquiring units.

The impact of this was only observed in CI and is not critical (some uploads are retried).

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
* none

Lazin committed:
archival: Add 'get_rw_fence' method

Currently, the read-write fence is created by the 'ntp_archiver'
for every operation. The process is manual. The fence is consturcted
using the in-memory state of the STM and feature table.

This commit improves this by moving this logic into a dedicated method
called 'get_rw_fence'. The method constructs the fence and checks if the
'_mutex' is acquired. We don't want to run more than one concurrent
fibers. When this happens both/all fibers may acquire the same fence
value and only one of the fibers will be able to commit the changes
happily.

The 'ntp_archiver' is single threaded. No concurrency is allowed there
(with the exception of data movement). So the invariant is that only one
fence can be generated at a time. The '_mutex' check ensures that this
is the case.

Some unit-tests required change.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
archival: Emit fence in reset-scrubbing-metadata

The method is invoked via the admin api. If it's invoked in parallel
with the upload the upload fails because there is an additional command
applied to the STM state. This makes the fence of the in-flight upload
invalid.

The commit adds a fence to the replicated scribber command and also
acquires units from the '_mutex' to avoid concurrency.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

WillemKauf commented:
Which CI failures/log lines does this PR fix? Is this related to the failed segment index uploads we have discussed offline? Please add the exact context to cover letter

Lazin commented:
> Which CI failures/log lines does this PR fix? Is this related to the failed segment index uploads we have discussed offline? 

This is something I discovered while fixing another issue. The uploads are failing because of the unaccounted concurrency but the test passes.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--675eb0eb-7cba-4cf1-b18d-fb3149d5d740-->
<!--TEST RESULTS-->
<!--948eb2be-49aa-44a4-9ed5-17165244e8b4-->
<!--TEST RESULTS-->
<!--dd1089ae-c1ec-45a6-ac07-1cdfd454f814-->
<!--TEST RESULTS-->
<!--9cd2e044-0da3-4e1a-95fd-0b03d3aef3e0-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64295">build#64295</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/64295#01961730-b8f0-4acc-bbd0-d9463dcfa691 | FLAKY | 1/2 |
| rptest.tests.cloud_retention_test.CloudRetentionTest.test_cloud_retention.max_consume_rate_mb=None.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/64295#0196178d-f1f8-411e-92d3-2b0ffffad264 | FLAKY | 1/2 |
| rptest.tests.follower_fetching_test.FollowerFetchingTest.test_basic_follower_fetching.read_from_object_store=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64295#01961779-de84-482d-9545-61a497fe3feb | FLAKY | 1/2 |
| rptest.tests.read_replica_e2e_test.TestReadReplicaService.test_writes_forbidden.partition_count=10.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/64295#01961779-de84-482d-9545-61a497fe3feb | FLAKY | 1/2 |
| rptest.transactions.tx_streams_test.TransactionsStreamsTest.consumer_offsets_retention_test | ducktape | https://buildkite.com/redpanda/redpanda/builds/64295#01961779-de84-482d-9545-61a497fe3feb | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64315">build#64315</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/64315#01961a26-6da2-4c62-ae0d-6bc76094c478 | FLAKY | 1/2 |
| rptest.tests.e2e_shadow_indexing_test.EndToEndShadowIndexingTestCompactedTopic.test_write.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/64315#01961a26-6da4-4cf1-be31-7f35642d7f66 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64482">build#64482</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| cloud_storage_rpfixture.cloud_storage_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/64482#01963494-5142-48d0-bbe9-11e384c80b6e | FAIL | 0/2 |
| cloud_storage_rpfixture.cloud_storage_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/64482#01963494-5143-4120-aacf-97d69bdb9744 | FAIL | 0/2 |
| gtest_cloud_storage_rpfixture.gtest_cloud_storage_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/64482#01963494-5142-48d0-bbe9-11e384c80b6e | FAIL | 0/2 |
| gtest_cloud_storage_rpfixture.gtest_cloud_storage_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/64482#01963494-5143-4120-aacf-97d69bdb9744 | FAIL | 0/2 |
| test_archival_service_rpfixture.test_archival_service_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/64482#01963494-5142-48d0-bbe9-11e384c80b6e | FAIL | 0/2 |
| test_archival_service_rpfixture.test_archival_service_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/64482#01963494-5143-4120-aacf-97d69bdb9744 | FAIL | 0/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64543">build#64543</a></summary>


| test_id                                                     | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| cloud_storage_rpfixture.cloud_storage_rpfixture             | unit        | https://buildkite.com/redpanda/redpanda/builds/64543#019638b9-e560-4cf6-aad8-99959e56b25c | FAIL          | 0/2      |
| cloud_storage_rpfixture.cloud_storage_rpfixture             | unit        | https://buildkite.com/redpanda/redpanda/builds/64543#019638b9-e561-4d9e-9970-83a002d98533 | FAIL          | 0/2      |
| gtest_cloud_storage_rpfixture.gtest_cloud_storage_rpfixture | unit        | https://buildkite.com/redpanda/redpanda/builds/64543#019638b9-e560-4cf6-aad8-99959e56b25c | FAIL          | 0/2      |
| gtest_cloud_storage_rpfixture.gtest_cloud_storage_rpfixture | unit        | https://buildkite.com/redpanda/redpanda/builds/64543#019638b9-e561-4d9e-9970-83a002d98533 | FAIL          | 0/2      |
</details>
<!--/body-->

oleiman commented on src/v/cluster/archival/ntp_archiver_service.cc:
```diff
@@ -2454,6 +2441,28 @@ std::ostream& operator<<(std::ostream& os, wait_result fr) {
+          features::feature::cloud_storage_metadata_rw_fence);
+
+    if (emit_read_write_fence && _mutex.available_units() > 0) {
+        throw std::runtime_error(fmt::format(
+          "[{}] Concurrency violation. Fence acquired without holding a mutex.",
+          _ntp.path()));
```
is this actually an exceptional situation at runtime, or would it always be a logic error? i.e. should it be a `vassert`?

nvartolomei replied:
I'm not familiar with Evgeny's reasoning but sometimes there are cases like this when it is hard to prove at test time that an invariant is held or to inform users of that (intricate) API what pre-conditions must be held.

In these cases throwing might be preferable as the system in most of the cases will continue functioning (a fact that is half proved by all our tests passing. for the other half we need to prove that all internal invariants are still held i.e. exceptional safety... that might be harder and i just confused myself with this rule of thumb too but I'm still inclined to throwing rather than vassert especially with backport) instead of shutting it down completely.

WillemKauf replied:
I think this could be a good use for a `dassert()`, if we felt strongly that this could be an unintended state rather than an exceptional one but didn't want to outright `vassert()`.

Lazin replied:
I changed this to `vassert`. Originally, I didn't use it to avoid potential crashloop but it looks like it's easy enough to check all uses of this method.

nvartolomei commented:
> @Lazin: "... The uploads are failing because of the unaccounted concurrency but the test passes ..."

Is it possible to add a new test to catch this?

Edit after reviewing the change: seems unlikely. lgtm.

nvartolomei reviewed (dismissed):
@Lazin the change is lgtm but while reviewing I noticed that `apply_spillover` does not use a fence. Is this a bug?

Lazin committed:
archival: Emit rw-fence in apply_spillover

The spillover is applied under the mutex but the fence is not applied.
This commit fixes this problem.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

WillemKauf reviewed (dismissed):
lgtm

Lazin commented:
> I noticed that apply_spillover does not use a fence. Is this a bug?

It's a bug. Fixed that.

Lazin commented:
> Is it possible to add a new test to catch this?

I don't think so. It's a race condition and the code will now trigger an assertion in this case.

Lazin commented:
Fixed CI failure caused by the last commit.

oleiman reviewed (dismissed):
seems legit 👍
